subsys/bluetooth/services/ias.c

```diff
@@ -0,0 +1,137 @@
+/** @file
+ *  @brief Immediate Alert Service implementation
+ */
+
+/*
+ * Copyright (c) 2022 Codecoup
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <init.h>
+#include <stdint.h>
+#include <zephyr.h>
+#include <net/buf.h>
+#include <logging/log.h>
+#include <bluetooth/gatt.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/services/ias.h>
+
+#define LOG_LEVEL CONFIG_BT_IAS_LOG_LEVEL
+LOG_MODULE_REGISTER(ias);
+
+#define BT_IAS_ALERT_MAX_LEN 1
+
+#if defined(CONFIG_BT_IAS_DEFAULT_PERM_W_AUTHEN)
+#define IAS_GATT_PERM_DEFAULT BT_GATT_PERM_WRITE_AUTHEN
+#elif defined(CONFIG_BT_IAS_DEFAULT_PERM_W_ENCRYPT)
+#define IAS_GATT_PERM_DEFAULT CONFIG_BT_IAS_DEFAULT_PERM_W_ENCRYPT
+#else
+#define IAS_GATT_PERM_DEFAULT BT_GATT_PERM_WRITE
+#endif
+
+typedef enum __packed {
+	BT_IAS_ALERT_LVL_NO_ALERT,
+	BT_IAS_ALERT_LVL_MILD_ALERT,
+	BT_IAS_ALERT_LVL_HIGH_ALERT,
+} bt_ias_alert_lvl_t;
+
+struct alerting_device {
+	bt_ias_alert_lvl_t alert_level;
+};
+
+static struct alerting_device devices[CONFIG_BT_MAX_CONN];
```

Comment:
Seems like a good approach :) The implementation seems to require a bit more updating before this is fully utilized, so I'll let you finish up before reviewing again

Reply:
Please elaborate on what do you mean by "updating"? :)

Reply:
The alert callbacks should be updated to also provide which connection is sending the alert, since there with multiple clients no longer is a single alert value. We should let it be up to the application to determiner how to properly handle the alert level per connection basis, instead of just returning the highest alert level.

Reply:
@Thalley I think this is overkill. Unless we expect applications to build its own priority table for bonded devices. e.g. lets ignore my watch alerts, and my alarm clock is highest prio... need to say it sounds weird.

In my head IAS shall be Easy peasy - device is in alert mode as long as any of the device is alerting, and with the highiest mode that any of the alerting device have.
I would not complicate it :)

Reply:
If we want to keep it that simple, we should have a single alert level instead. 

I don't like this mix-match between supporting multiple clients in the service without providing that to the upper layers. The only thing we achieve by doing that is throwing away information. 

I suggest that we
1) Support multiple clients and provide that information to the upper layers (as easy as just providing the `conn` pointer). The application can then decide how to deal with this (i.e. it doesn't need to create a table for connected devices, but can if it so wants to). 
2) Keep the alert level as a singleton as the original implementation was (the service specification is unclear about this IMO). 

Personally I'm in favor of 1, but will also accept 2.

Reply:
Ad1: I am afraid each application will have to implement it in the same way. Why not to handle it in the stack? I'm in favor in current API design which is kind of simple one. If you have strong opinion on having `conn` in callback which will give application opportunity to play with it - well that is the easiest thing to do.  

Ad2. not really sure what you mean by "alert level as a singleton".  
The `static struct alerting_device devices[CONFIG_BT_MAX_CONN];` is needed to track disconnection of the device. AS per spec if device disconnects we should clear alert from that device. Since connected devices does not know about each other, we should not clear alert from other device. So I suggest to keep that if we go with this option.

Personally I'm in favor of keeping as it is, but if you have strong option to go for 1) we will do it - but then it means, HAP will have to implement this logic by its own.

Reply:
> The static struct alerting_device devices[CONFIG_BT_MAX_CONN]; is needed to track disconnection of the device. 

That is true, but it is currently also being used to store alert levels for all connections.

If device_A writes a high alert then the application is notified. If device_B writes a medium alert, then the application is not notified. If device_A disconnects, the the application is suddenly notified that the alert changed from High to Medium, and it may only know that device_A is disconnected, but it cannot tell if the change in alert comes from a previous alert or a new alert. 



In the case that two devices device_A and device_B connects, and device_A writes a high alert (using previous example), and device_B writes a medium alert, then the application is only notified about the medium alert. 
When the application gets a high alert, it probably means that it (or the user) needs to do something. After the application/user has done that, it will probably call `bt_ias_local_alert_stop`, but that will also stop the medium alert from device_B, which may not have been handled as it was another alert level. 



If you are mostly keen on keep the API simple (i.e. omitting which device writes the alert), can I suggest the alternative of keeping the alert level in the application instead? Then at least the application get all the alert writes, and can handle it however it wants if it gets both a high and a medium alert. 

I highly dislike the idea the idea of the stack simply silently ignoring or hiding information from the upper layers - Especially in this case where the alert is the one and only thing this service provides.

Reply:
Perhaps input from @MariuszSkamra or @asbjornsabo would be helpful :)

Reply:
> If device_A writes a high alert then the application is notified. If device_B writes a medium alert, then the application is not notified. If device_A disconnects, the the application is suddenly notified that the alert changed from High to Medium, and it may only know that device_A is disconnected, but it cannot tell if the change in alert comes from a previous alert or a new alert

The job to do for the device producing alerts it just produce the alert when needed. Medium alert could be buzzzer, high could be buzzer and sound or whatever. If you take for the moment perspective of such a device, why it should care who is alerting - there is a job to do which is alert on the request and if one of the device request High alert there is not much to do but follow that. If that for some reason goes down to Medium, then device will go for that. Bluetooth stack cares if there is any device requesting alerts and as long somebody request, the highest alert requested should be executed. I cannot really see any other policy than that. e.g. I cannot imagine what are other options to do for the device which gets two different alerts request. 

> I highly dislike the idea the idea of the stack simply silently ignoring or hiding information from the upper layers - Especially in this case where the alert is the one and only thing this service provides.

well, each service expose things in API things which matters for the user. In my opinion, current API and service implementation does exactly this.

Reply:
@rymanluk @Thalley maybe we could have something similar to struct `_bt_gatt_ccc`. There are 2 types of callbacks: One `alert_level_changed` which is global, highest value of level and 2nd `alert_level_write` which is per conn.

Reply:
@MariuszSkamra thanks! Sounds good. @szymon-czapracki @Thalley  what do you think?

Reply:
I am not a fan of that approach. In fact, that approach for the CCC has issue that we are trying to somewhat solve in https://github.com/zephyrproject-rtos/zephyr/pull/44136. 

Having two callbacks does not IMO make anything any simpler, but it will provide the functionality that I request. 

So my preference list would be
1) Add `struct bt_conn` to the callbacks, and let the application decide how to handle the alarms (which, by spec definition is implementation specific). 
2) Add both the connectionless callbacks and the `struct bt_conn` callbacks, and let the application decide whether to handle it itself, or use the "get the highest alert level" approach
3) Keep as is

Reply:
Actually, since I have not (and can't really at the moment) provide any specific use case for my suggestion, I will approve this approach. I'll bring it up again if I think of a use case.

Reply:
I've talked to the Bluetooth specification guru at our office. His opinion is that the behavior should be left to a higher level protocol. It would be natural to have an Immediate Alert Profile that defines behavior, but this does not exist. It's possible the intention is that this service should be referred to by more specific profiles, like an imagined "Fire Indicator Profile".

We can of course make choices for our applications / higher layers, but we should be aware that our API may be less general than IAS as it's specified and we should document that and we may even want to define a profile.

Some notes about the spec:
  - Interestingly, IAS requires the alarm to continue *at least* until the specified conditions occur. It is unspecified and therefore allowed for the alarm to continue after the conditions occur.
  - IAS does not specify the behavior for multiple devices. It is again up to the application / profile. Ideally, enough information is presented to upper layers to be as general as possible.

Reply:
The IAS spec is fairly old, and that's noticeable when reading it. Newer GATT specs (such as the LE Audio specs) are more specific and more well written IMO. 

We could consider opening an erratum for IAS to add some note or recommendation or how it should work with multiple devices.

Reply:
@alwa-nordic glad that Bluetooth specification guru found a time to look at this problem. Before we do it, can we get the use case which would be different from the policy we described here many times? ;)